File: 02MultipleOutputs/toBeDifferentiated.f
C Small example to differentiate many dependent variables
C  with respect to a single independent variable. 
C  i.e. f: R^1 \to R^n 

      PROGRAM MAIN 
        IMPLICIT NONE
        REAL*8 innum, outcos, outsin, rad 

        REAL*8 getprodsum

        innum = 1.23

        rad = GETPRODSUM(innum, outcos, outsin)

        print *, "Radius = ", rad, " -- cos = ", outcos,  
     + " - sin = ", outsin
      END PROGRAM MAIN

      FUNCTION GETPRODSUM(in, outcos, outsin)
        IMPLICIT NONE
        REAL*8 in, outcos, outsin
        REAL*8 tmp 
        
        REAL*8 getprodsum

        tmp = in*in/2.0+in+1.0

        outcos = cos(tmp);
        outsin = sin(tmp);

        getprodsum = outcos*outcos+outsin*outsin
      END FUNCTION GETPRODSUM

File: 02MultipleOutputs/toBeDifferentiated.c
/* Small example to differentiate many dependent variables
 * with respect to a single independent variable.
 * i.e. f: R^1 \to R^n */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

double get_radius(double in, double* outcos, double* outsin) {
    double tmp = in*in/2.0+in+1.0;

    *outcos = cos(tmp);
    *outsin = sin(tmp);

    return (*outcos)*(*outcos)+(*outsin)*(*outsin);
}

int main(int argc, char* argv[]) {

    double innum;
    double outcos;
    double outsin;
    double rad;

    if (argc != 2)
    {
        printf("ERROR: some angle should be provided as first argument!\n");
        return EXIT_FAILURE;
    }

    innum = atof(argv[1]);

    rad = get_radius(innum, &outcos, &outsin);

    printf("Radius = %f \t -- \t cos = %f - sin = %f\n", rad, outcos, outsin);
    return EXIT_SUCCESS;
}
